Source/JavaScriptCore/wasm/WasmTypeDefinition.h

```diff
@@ -94,8 +94,127 @@ enum Mutability : uint8_t {
     Immutable = 0
 };
 
-struct FieldType {
-    Type type;
+inline const char* makeString(PackedType packedType)
+{
+    switch (packedType) {
+    case PackedType::I8:
+        return "i8";
+    case PackedType::I16:
+        return "i16";
+    }
+    ASSERT_NOT_REACHED();
+}
+
+struct StorageType {
+public:
+    template <typename T>
+    bool is() const { return std::holds_alternative<T>(m_storageType); }
+
+    template <typename T>
+    const T* as() const { ASSERT(is<T>()); return std::get_if<T>(&m_storageType); }
+
+    StorageType() {
+        // FIXME: ???
+        m_storageType = std::variant<Type, PackedType>();
+    }
+
+    StorageType(Type t) {
+        m_storageType = std::variant<Type, PackedType>(t);
+    }
+
+    StorageType(PackedType t) {
+        m_storageType = std::variant<Type, PackedType>(t);
+    }
+
+    // If this is a type, returns itself; if it's a packed type, returns i32
+    Type unpacked() const {
+        if (is<Type>()) {
+            return *as<Type>();
+        }
+        return Types::I32;
+    }
+
+    size_t elementSize() const {
+        if (is<Type>()) {
+            switch (as<Type>()->kind) {
+            case Wasm::TypeKind::I32:
+            case Wasm::TypeKind::F32:
+                return sizeof(uint32_t);
+            default:
+                return sizeof(uint64_t);
+            }
+        }
+        switch (*as<PackedType>()) {
+        case PackedType::I8:
+            return sizeof(uint8_t);
+        case PackedType::I16:
+            return sizeof(uint16_t);
+        }
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    bool operator==(const StorageType& rhs) const {
+        if (rhs.is<PackedType>()) {
+            return (is<PackedType>() && as<PackedType>() == rhs.as<PackedType>());
+        }
+        if (!is<Type>()) {
+            return false;
+        }
+        return(as<Type>() == rhs.as<Type>());
+    }
+    bool operator!=(const StorageType& rhs) const { return !(*this == rhs); };
+
+    auto hash(uint32_t start) const {
+        unsigned accumulator = start;
+        uint8_t ty;
+        if (is<Type>()) {
+            ty = static_cast<uint8_t>(as<Type>()->kind);
+        } else {
+            switch (*as<PackedType>()) {
+            case PackedType::I8:
+                ty = JSC::Wasm::numTypes;
+                break;
+            case PackedType::I16:
+                ty = JSC::Wasm::numTypes + 1;
+                break;
+            }
+        }
+        accumulator = WTF::pairIntHash(accumulator, WTF::IntHash<uint8_t>::hash(ty));
+        accumulator = WTF::pairIntHash(accumulator, WTF::IntHash<uint8_t>::hash(static_cast<uint8_t>(nullable())));
+        accumulator = WTF::pairIntHash(accumulator, WTF::IntHash<TypeIndex>::hash(index()));
+        return accumulator;
+    }
+
+    bool nullable() const {
```

Comment:
If `nullable()` is only used in the hashing function above, you can likely delete it and the use above. There was a patch that landed somewhat recently that removed the nullability flag in the type representation (since `Ref` and `RefNull` type codes encode all the necessary information, and that's always hashed) so it's not needed to hash that.